lib/gems/pending/util/miq_file_storage.rb

```diff
@@ -0,0 +1,203 @@
+# This class is meant to be a abstract interface for defining a file_storage
+# class.
+#
+# The storage class can either be of a type of "object storage", which includes:
+# * protocols like FTP
+# * document storage like s3 and OpenStack's Swift
+#
+# And mountable filesystems like:
+# * NFS
+# * SMB
+#
+# The class is meant to allow a shared interface for working with these
+# different forms of file storage, while maintaining their differences in
+# implementation where necessary.  Connection will be handled separately by the
+# subclasses, but they must conform to the top level interface.
+#
+class MiqFileStorage
+  def self.with_interface_class(opts)
+    klass = fetch_interface_class(opts)
+    block_given? ? yield(klass) : klass
+  end
+
+  def self.fetch_interface_class(opts)
+    return nil unless opts[:uri]
+
+    require 'uri'
+    scheme, _ = URI.split(URI::DEFAULT_PARSER.escape(opts[:uri]))
+    klass = storage_interface_classes[scheme]
+    klass.new_with_opts(opts) if klass
+  end
+  private_class_method :fetch_interface_class
+
+  def self.storage_interface_classes
+    @storage_interface_classes ||= Interface.descendants.each_with_object({}) do |klass, memo|
+      memo[klass.uri_scheme] = klass if klass.uri_scheme
+    end
+  end
+
+  class Interface
+    BYTE_HASH_MATCH = /^(?<BYTE_NUM>\d+(\.\d+)?)\s*(?<BYTE_QUALIFIER>K|M|G)?$/i
+    BYTE_HASH       = {
+      "k" => 1.kilobyte,
+      "m" => 1.megabyte,
+      "g" => 1.gigabyte
+    }.freeze
+
+    attr_reader :remote_file_path, :byte_count, :source_input, :input_writer
+
+    def self.new_with_opts(opts) # rubocop:disable Lint/UnusedMethodArgument
+      raise NotImplementedError, "#{name}.new_with_opts is not defined"
+    end
+
+    def self.uri_scheme
+      nil
```

Comment:
Related to my previous comment. Should this also raise `NotImplementedError` or are we expecting some classes to leave this as `nil`?

Reply:
Related to my [previous answer](https://github.com/ManageIQ/manageiq-gems-pending/pull/361#discussion_r216506316) (:trollface:), if this isn't implemented, this won't get added to the lookup hash.